internal/validators/namespace.go

```diff
@@ -203,28 +218,41 @@ func (v *Namespace) illegalCascadingDeletion(ns *forest.Namespace) admission.Res
 // decodeRequest gets the information we care about into a simple struct that's easy to both a) use
 // and b) factor out in unit tests.
 func (v *Namespace) decodeRequest(log logr.Logger, in admission.Request) (*nsRequest, error) {
```

Comment:
I think the changes (and probably the overall complexity) of this function could be reduced if you keep the existing structure. E.g.:

* if op is delete, decode `in.OldObject.Raw` into `ns`
* else, decode the request into `ns`, and decode `in.OldObject.Raw` into `oldNS` if it's non-nil

We don't support K8s 1.14 anymore so you can take out the special-case code for that.

This preserves all existing behaviour (and code) and _only_ modifies the code when oldNS is relevant. Even most of your changes to the existing tests would go away.